File: sample_source_code/csharp/src/Cat.cs
namespace Embold.Sample;

/// <summary>Concrete Cat class with lifespan estimation.</summary>
public sealed class Cat : Animal
{
    public Cat(string name, int age, bool isIndoor, double weight)
        : base(name, age, "Felis catus", weight)
    {
        IsIndoor = isIndoor;
    }

    public bool IsIndoor { get; }

    public override string Sound() => "Meow";
    public override void   Accept(IVisitor visitor) => visitor.Visit(this);

    /// <summary>Returns expected lifespan adjusted for environment and weight.</summary>
    public int ExpectedLifespan()
    {
        int    baseLine   = IsIndoor ? 15 : 10;
        double deviation  = Math.Abs(Weight - 4.5) / 4.5;
        int    adjustment = (int)MathUtils.LinearDecay(deviation, 1.0, 5);
        return Math.Max(5, baseLine - adjustment);
    }
}

File: sample_source_code/CPP/src/Cat.cpp
#include "Cat.h"
#include "MathUtils.h"
#include <cmath>
#include <algorithm>

Cat::Cat(const std::string& name, int age, bool isIndoor, double weight)
    : Animal(name, age, "Felis catus", weight), isIndoor_(isIndoor) {}

bool        Cat::isIndoor() const { return isIndoor_; }
std::string Cat::sound()    const { return "Meow"; }

void Cat::accept(IVisitor& visitor) { visitor.visit(*this); }

int Cat::expectedLifespan() const {
    int    base      = isIndoor_ ? 15 : 10;
    double deviation = std::abs(weight_ - 4.5) / 4.5;
    int    adjustment = static_cast<int>(MathUtils::linearDecay(deviation, 1.0, 5.0));
    return std::max(5, base - adjustment);
}
